Replace string[] with int[] for dimensions

navis opened the pull request:
Currently dimdim conveys string[] which makes need for binary search for each values in making index file. But with int[], we can easily translate transient ids into index ids and also need not use interner for easing checking for equality of two values.

binlijin commented:
:+1:

drcrallen commented:
Hi Navis, thanks for the PR. Do you have performance numbers for this change?

navis commented:
@drcrallen I don't have any numbers yet. Is there any test which can check performance of generating druid index or proper test data to be used for the purpose in druid?

navis commented:
Wishfully this can be a start for implementing non-string type dimesion in druid.

navis commented:
@KurtYoung

> I think retVal should be sorted before return

I forgot to ask a question on this. Why it should be sorted? I think multi-vals dimension is an array not a set. For arrays, should {a, b} be different from {b, a} ?

navis commented:
Oh, I've got numbers just now. I've used lineitem file for tpch test (the biggest tsv format file in my notebook), which has 6M rows with 16 dimensions ranged from very high cardinality (comments) to very low (line status). Used orderkey as a timestamp and ran with maximum 50000 rows, making 12 rounds. I've also modified RealtimePlumber not to use thread-pool for persisting (I've got OOM without this).  

current (indexing time, persisting time)
6246, 45304
5638, 44682
6205, 44874
6180, 43834
5746, 42781
6083, 42945
5985, 46322
5590, 44279
5701, 41627
6024, 43390
5590, 43473
6045, 44472

with patch
5027, 36039
4934, 36111
5078, 35252
5016, 33591
5073, 34376
4995, 36473
4912, 34717
5223, 34209
5102, 35388
5067, 36582
4986, 34888
4390, 35132

I expected increase on indexing time and decrease on persisting time but both decreased. I've not checked any validity of the result, yet. I'll do that tomorrow (Time to go home).

xvrl commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -328,13 +329,14 @@ public IncrementalIndex(
       columnCapabilities.put(entry.getKey(), capabilities);
     }
-    this.dimValues = new DimensionHolder();
+    this.dimValues = new DimensionHolder(32 + (dimensions.size() + metrics.length) << 2);
```
can you explain this?

navis replied:
It's my laziness, wanting the array final. I'll check performance difference with proper increasing-array semantic.

xvrl commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -660,35 +665,33 @@ public DateTime getMaxIngestedEventTime()
+    private final DimDim[] dimensions;
 
-    DimensionHolder()
+    DimensionHolder(int maxDimensions)
```
what happens when we exceed maxDimensions? How will it work with schema-less indexing?

navis replied:
IIOB, maybe? commented above about the plan.

navis commented:
Addressed comments and new numbers. Can see some fluctuation of indexing time.

4967, 36014
5320, 35638
5244, 36398
5344, 35869
4835, 34723
5272, 35275
4329, 33935
4725, 34961
5178, 34585
4918, 35582
5529, 35182
5237, 35334

navis commented:
Is there anything I can do for this to be merged?

KurtYoung commented:
I think whether multi value should be sorted should be discussed. 
In some scenario multi value sorted can also make scene. For a multi value dimension expressing the properties of something, like a=b,c=d，the separator is ",". In this case, I'd rather treat "a=b,c=d" and "c=d,a=b" as the same value

navis commented:
@KurtYoung Thanks for the explanation. I think this should be optionized in somewhere. For now, I'll restore sorting part. Let's do the optionizing in following issue.

navis commented:
Met #2128 and #2115. Rerunning test.

fjy commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -658,37 +671,42 @@ public DateTime getMaxIngestedEventTime()
+    private DimensionHolder(int numDimension)
     {
-      dimensions = Maps.newConcurrentMap();
+      dimensions = new DimDim[Math.max(32, numDimension)];
```
Why is there a minimum of 32 entries created?

fjy replied:
also, the previous data structure was a concurrent map, how is locking done over this array?

navis replied:
> Why is there a minimum of 32 entries created?

It would be any number which is >=1. Just choosed blindly. 

> also, the previous data structure was a concurrent map, how is locking done over this array?

Put() method is always call in lock (`dimensionOrder`), but missed lock for get(). I'll add one for that.

fjy commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -658,37 +671,42 @@ public DateTime getMaxIngestedEventTime()
+    private DimDim[] dimensions(int expectedIndex)
+    {
+      if (expectedIndex >= dimensions.length) {
+        dimensions = Arrays.copyOf(dimensions, dimensions.length + Math.min(1024, dimensions.length));
```
where does 1024 come from? why this number?

navis replied:
Again, blindly choosen. Would it be better to just double it?

fjy commented on processing/src/test/java/io/druid/query/groupby/GroupByQueryRunnerTest.java:
```diff
@@ -831,7 +831,7 @@ public String apply(String dimValue)
 
     List<Row> expectedResults = Arrays.asList(
-        GroupByQueryRunnerTestHelper.createExpectedRow("2011-04-01", "alias", "", "rows", 3L, "idx", 2870L),
+        GroupByQueryRunnerTestHelper.createExpectedRow("2011-04-01", "alias", null, "rows", 3L, "idx", 2870L),
```
this is a change in behavior and one we have to be careful about. I think it would be nice if we can preserve that empty strings are returned in place of nulls for results.

navis replied:
IncrementalIndex regards null and empty string altogether as empty string and stores them like that in fact map. But when retrieving values, which uses dimdim.getValue(), it converts empty strings to nulls. So we can say most of queries returns null for empty string. 

But for group-by queries, it makes IncrementalIndex in the end of query and retrieves values from the fact map in it directly(in `IncrementablIndex#iterableWithPostAggregations`), which makes empty strings returned as-is.

With this patch, there is no direct way to retrieve values from fact map because now we stores int[] in it rather than string[]. So I can say that it's more consistent than before. We can simply revert nulls to empty strings again in the method to make the test pass but I think it's not a good idea. Could you make decision on this, @fjy?

navis replied:
@fjy I think this is the sole comment not addressed yet. Could you check this?

fjy replied:
@navis I think we should be consistent in returning null or empty string to clients. I believe right now Druid will always try to return empty string. Can we do a null to empty string conversion within the groupBy code?

xvrl replied:
in the index we treat null and empty string the same, storing it as empty string, but to the user, Druid always returns null

navis replied:
I think you guys didn't read my comment. But ok, I've added `nullToEmpty` to it and rolledback the tests.

fjy commented on processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java:
```diff
@@ -245,92 +243,97 @@ public Object getMetricObjectValue(int rowOffset, int aggOffset)
     }
 
-    public int getSortedId(String value)
+    private synchronized void addNewValue(int index, String value)
```
instead of synchronizing on this, can we synchronize on the code using a lock?

navis replied:
done.

fjy commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -658,37 +671,42 @@ public DateTime getMaxIngestedEventTime()
 
-    DimDim add(String dimension)
+    // called in lock
+    private void put(String dimension, int index)
```
is there any locking over this put method?

navis replied:
done.

fjy commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -757,38 +772,48 @@ public int add(String value)
+    public DimLookup sort()
     {
-      return delegate.getSortedId(Strings.nullToEmpty(value));
+      return new NullValueConverterDimLookup(delegate.sort());
```
i think it is nice to preserve the behavior here that nulls are returned as empty strings

in general in Druid, nulls and empty strings are treated the same

navis replied:
IncrementalIndex regards null and empty string altogether as empty string and stores them like that in fact map. But when retrieving values, which uses dimdim.getValue(), it converts empty strings to nulls. So we can say most of queries returns null for empty string. 

But for group-by queries, it makes IncrementalIndex in the end of query and retrieves values from the fact map in it directly(in `IncrementablIndex#iterableWithPostAggregations`), which makes empty strings returned as-is.

With this patch, there is no direct way to retrieve values from fact map because now we stores int[] in it rather than string[]. So I can say that it's more consistent than before. We can simply revert nulls to empty strings again in the method to make the test pass but I think it's not a good idea. Could you make decision on this, @fjy?

navis replied:
There is no `getSortedId()` anymore which converts value into index. Instead we translates int[] directly.

fjy commented:
@navis i still have some outstanding comments before this can be merged

jon-wei commented:
@navis I noticed the changes here conflict with the #2169, what are your future plans for this PR?

navis commented:
@jon-wei Merge any non-commited PR on any commited PR? Don't have any special plan, but I expect some following works can be started from this PR.

drcrallen commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -504,26 +508,28 @@ private long getMaxTimeMillis()
+      return new int[]{dimLookup.add(dimValues.get(0))};
+    }
+
+    // todo: make option for this
```
TODO

navis replied:
Thinking of doing it later. We need some dimension spec (not capability) which can be used for describing it in more detail (something like types, options, comparators, null-handling, etc.) which is currently not available.

fjy replied:
@navis can we file an issue versus having a TODO in the code?

navis replied:
Ok, I'll make one.

jon-wei commented:
@navis Cool, I wanted to know if you were going to update this PR for IncrementalIndex after #2169's merge, just something for me to keep in mind. I am working on a patch that loosens the types for dimensions in IncrementalIndex that I plan on submitting a PR for in the next few days.

navis commented:
@jon-wei Good to know. All theses refactoring are intended for that kind of issues because we also need numeric type dimensions. I'll rebase this.

navis commented:
Rebased on trunk and squashed commit lists

nishantmonu51 commented:
still showing conflicts, can you rebase again and fix conflicts  ?

navis commented:
Rebased again

navis commented:
rebased on trunk couple of times

fjy commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -512,26 +516,28 @@ private long getMaxTimeMillis()
+
+    // todo: make option for this
+    String[] dimArray = dimValues.toArray(new String[dimValues.size()]);
+    Arrays.sort(dimArray);
```
why is this sort needed?

fjy commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -512,26 +516,28 @@ private long getMaxTimeMillis()
-      retVal[count] = canonicalDimValue;
-      count++;
+    for (String dimValue : dimArray) {
+      retVal[count++] = dimLookup.add(dimValue);
```
why did we remove the optimization here?

navis replied:
Now we stores int[] array instead of String[], it's not need to be canonicalized.

himanshug replied:
nit: `for (int count = 0; count < dimArray.length; count++) { .. }` would be more straight forward.

navis replied:
It's a kind of 'keeping the old codes not broken'. I'll replace that.

fjy commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -694,15 +714,28 @@ public Row apply(final Map.Entry<TimeAndDims, Integer> input)
+                  }
+                  String dimensionName = dimensionDesc.getName();
+                  if (dim == null || dim.length == 0) {
+                    theVals.put(dimensionName, null);
```
this is changing behavior from the logic that was there before. Why?

navis replied:
Because It's now int[], we need convert it to String[]

fjy commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -694,15 +714,28 @@ public Row apply(final Map.Entry<TimeAndDims, Integer> input)
+                    theVals.put(dimensionName, dimensionDesc.getValues().getValue(dim[0]));
                   } else {
-                    theVals.put(dimensions.get(i), null);
+                    String[] multiple = new String[dim.length];
```
can we call this dimStringValue?

navis replied:
sure

fjy commented on processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java:
```diff
@@ -275,92 +274,111 @@ public void close()
-    public OnHeapDimDim()
+    // for faster access
+    @GuardedBy("valueToId")
+    private String[] idToValue = new String[32];
```
how did u decide this size?

fjy replied:
how much of a perf diff does this actually make over a list?

navis replied:
Minimal. Changed to List

fjy commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -730,6 +763,45 @@ public DateTime getMaxIngestedEventTime()
+    {
+      synchronized (lock) {
+        if (expectedIndex >= dimensions.length) {
+          dimensions = Arrays.copyOf(dimensions, dimensions.length + Math.min(1024, dimensions.length));
```
can we just use a list instead of the explicit array copying?

navis replied:
Changed to List

fjy commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -730,6 +763,45 @@ public DateTime getMaxIngestedEventTime()
+
+    private DimensionHolder(int numDimension, Object lock)
+    {
+      this.dimensions = new DimDim[Math.max(32, numDimension)];
```
can we just use a list?

fjy replied:
there appears to be locking already in places where put is called.

jon-wei replied:
I'm also in favor of changing the arrays that require explicit resizing (DimensionHolder, OnHeapDimDim) to lists

navis replied:
Changed to List

fjy commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -730,6 +763,45 @@ public DateTime getMaxIngestedEventTime()
+      dimensions[index] = dimension.getValues();
+    }
+
+    private DimDim dimension(int index) {
```
call this getDimensions

fjy replied:
actually, this is getDimDims

navis replied:
Renamed to getDimValues

fjy commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -730,6 +763,45 @@ public DateTime getMaxIngestedEventTime()
+      this.lock = lock;
+    }
+
+    private void put(DimensionDesc dimension)
```
call dimension dimensionDesc to make the code easier to read

navis replied:
done

fjy commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -730,6 +763,45 @@ public DateTime getMaxIngestedEventTime()
+    private void put(DimensionDesc dimension)
+    {
+      final int index = dimension.getIndex();
+      final DimDim[] dimensions = dimensions(index);
```
dimDims not dimensions

navis replied:
renamed to dimValues

fjy commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -730,6 +763,45 @@ public DateTime getMaxIngestedEventTime()
+  {
+    @GuardedBy("lock")
+    private DimDim[] dimensions;
+    private final Object lock;
```
why this is lock needed?

fjy replied:
there appears to be locking already in places where put is called.

navis replied:
There might be concurrent access to facts by queries.

jon-wei commented on processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java:
```diff
@@ -275,92 +274,111 @@ public void close()
-    {
-      if (sortedVals == null) {
-        sortedVals = new String[falseIds.size()];
+  private static class OnHeapDimLookup implements DimLookup
```
Can you change DimLookup class name to something that indicates it's used for sorted values, e.g. SortedDimLookup

navis replied:
done

himanshug commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -512,26 +516,28 @@ private long getMaxTimeMillis()
-      if (!dimLookup.contains(null)) {
-        dimLookup.add(null);
-      }
+      dimLookup.add("");
```
why the behavior change? earlier it was dimLookup.add(null) and now changed to dimLookup.add("") .

navis replied:
Same with adding `null`. Changed like that.

himanshug commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -924,69 +1009,86 @@ long getTimestamp()
+
+  Comparator<TimeAndDims> dimsComparator()
+  {
+    return new TimeAndDimsComp();
```
TimeAndDimsComp can be a singleton, you can just have a static final INSTANCE in there

navis replied:
Because it's referencing a field(`dimValues`), it cannot be static.

navis commented:
Because there is so many conflicts, I think I should rebase this after #2094 is merged.

navis commented:
Rebase, anyway.

jon-wei commented:
:+1: looks good to me after addressing the Seeker renaming comments

himanshug commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -268,6 +270,7 @@ public int lookupId(String name)
   private final Map<String, MetricDesc> metricDescs;
   private final Map<String, DimensionDesc> dimensionDescs;
   private final Map<String, ColumnCapabilitiesImpl> columnCapabilities;
+  private final DimensionValueHolder dimValues;
```
it looks like that DimensionValueHolder is equivalent of `Collections.synchronizedList(Lists.newArrayListWithCapacity(Math.max(32, numDimension)))` , why introduce a whole new class?

himanshug replied:
check being done in DimensionValueHolder.put(..) can be moved to addNewDimension(..).

navis replied:
Almost two months had passed from the first patch and things are changed much in hereabout. I'll check if I can remove more codes in here.

navis replied:
removed DimensionValueHolder.

himanshug commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -512,26 +512,27 @@ private long getMaxTimeMillis()
+    }
+
+    String[] dimArray = dimValues.toArray(new String[dimValues.size()]);
+    Arrays.sort(dimArray);
```
why not `Collections.sort(dimValues)` and use dimValues instead of creating a new array with the copy?

navis replied:
Because the `dmValues` is from InputRow, which is dependent to user implementation, we don't know it's random accessible or not.

himanshug replied:
hmmm.... but, `get(int index)` is part of List interface contract. It will be very weird for someone to implement an InputRow which returns a dimension values List that does not implement get(int) . (in that case actually that InputRow implementation should be fixed).

navis replied:
Scratch (my) above comment. I'll be a problem when the list does not support set() method. For example, commonly used ones from ImmutableLists.of() or Lists.asList(), etc. are in this category.

himanshug commented:
@navis I'm trying to review this PR to get to closure. However, I'm not sure if I fully understood the advantage of storing int[] in TimeAndDims . It appears that the performance improvement is proposed by optimizing away the binary search in `OnHeapDimDim.getSortedId(String value)` which could be done by keeping a `Map<String,Integer> sortedDimValues` in OnHeapDimDim after the `sort()` (currently it keeps only sorted String[] which requires binary search)
is storing int[] in TimeAndDims necessary for proposed optimization?
also this PR introduces a shared lock object in DimDim and DimensionHolder which did not exist before, why is that needed now?

navis commented:
@himanshug

> It appears that the performance improvement is proposed by...

Yes, major points are from the direct conversion of id[] to index[]. And also we don't need interning Strings before adding to fact. 

> shared lock object in DimDim

By using int[] instead of string[], we need more frequent access to converting id back to value, which is `DimDim#getValue(int id)`. It was implemented by synchronized bitmap and I thought it just can be an List which seemed more faster than map. So additional lock is introduced to guard two collections.

> and DimensionHolder

removed that.

drcrallen commented:
@fjy / @himanshug any more comments here?

fjy commented:
:+1:

himanshug commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -812,7 +852,7 @@ public ColumnCapabilitiesImpl getCapabilities()
   static interface DimDim
   {
-    public String get(String value);
+    public String dimension();
```
this method does not seem to be used anywhere, do we need it?

navis replied:
removed that.

himanshug commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -553,6 +554,20 @@ private long getMaxTimeMillis()
     }
   }
 
+  public List<DimDim> getDimensionValues()
```
this is not used anywhere.

himanshug replied:
`getDimensionIndex(String dimension)` is also unused, i know this PR did not introduce that but can you pls remove that as well.

navis replied:
sure

himanshug replied:
@navis `getDimensionValues()` is still not removed.

navis replied:
ah, thanks.

himanshug commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -925,69 +978,86 @@ long getTimestamp()
+  private class TimeAndDimsComp implements Comparator<TimeAndDims>
+  {
+    @Override
+    public int compare(TimeAndDims lhs, TimeAndDims rhs)
```
this is non trivial, it will be nice to have some independent unit tests for the TimeAndDimsComp covering various corner cases.

navis replied:
ok

himanshug commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndexAdapter.java:
```diff
@@ -79,19 +107,19 @@ public IncrementalIndexAdapter(
-
-        if (bitmapIndexes == null || dims == null) {
+        DimensionIndexer indexer = indexers.get(dimension.getName());
+        if (indexer == null || dims == null) {
```
indexer can never be null, we just added it on line 112, no?

navis replied:
Null check was already there but I'll remove that.

navis commented:
Update on numbers. 500K Rows per each, msec

BEFORE

Adding to Heap | Walk(id to index) | Writing to Smoosh (Total)
-------------- | ------------------| ------------------
8177 | 15297 | 41502
6524 | 18959 | 42640
6433 | 14248 | 43228
7266 | 18568 | 43374
6516 | 14693 | 40294
6716 | 14435 | 41148
6996 | 17024 | 40764
6721 | 18761 | 44617
6934 | 18759 | 42688
7033 | 17098 | 42959
6594 | 15266 | 43943
6614 | 14166 | 39430


AFTER

Adding to Heap | Walk(id to index) | Writing to Smoosh (Total)
-------------- | ------------------| ------------------
6590 | 1989 | 10703
4454 | 2473 | 10721
4774 | 1520 | 9092
4713 | 1600 | 9459
4619 | 1610 | 9430
4583 | 1621 | 9072
4884 | 1534 | 9225
4545 | 1750 | 9542
4566 | 1840 | 9145

fjy commented:
@himanshug we good to go?

binlijin commented:
Also i think this will use more less memory.

himanshug commented:
@navis numbers look pretty good. I am :+1: after https://github.com/druid-io/druid/pull/2085/files#r51679583 is resolved and commits are squashed.

fjy commented:
@navis we can merge after @himanshug's last comment is addressed and commits are squashed
